packages/ckeditor5-watchdog/src/editorwatchdog.ts

```diff
@@ -323,6 +352,110 @@ export default class EditorWatchdog<TEditor extends Editor = Editor> extends Wat
 	}
 }
 
+class EditorWatchdogInitPlugin {
+	public editor: Editor;
+	private _data: EditorData;
+
+	constructor( editor: Editor ) {
+		this.editor = editor;
+
+		this._data = editor.config.get( '_watchdogInitialData' )!;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public init(): void {
+		this.editor.data.on( 'init', evt => {
+			evt.stop();
+
+			this.editor.model.enqueueChange( { isUndoable: true }, writer => {
+				this._restoreEditorData( writer );
+			} );
+
+			this.editor.data.fire<EditorReadyEvent>( 'ready' );
+
+			// Keep priority `'high' - 1` to be sure that RTC initialization will be first.
+		}, { priority: 1000 - 1 } );
+	}
+
+	private _createElement( writer: Writer, node: any ): Text | Element {
+		if ( 'name' in node ) {
+			// If child has name property, it is an Element.
+			const element = writer.createElement( node.name, node.attributes );
+
+			if ( node.children ) {
+				for ( const child of node.children ) {
+					element._appendChild( this._createElement( writer, child ) );
+				}
+			}
+
+			return element;
+		} else {
+			// Otherwise, it is a Text node.
+			return writer.createText( node.data, node.attributes );
+		}
+	}
+
+	/**
+	 * Restores the editor by setting all operations, attributes and markers.
+	 */
+	private _restoreEditorData( writer: Writer ): void {
+		const editor = this.editor!;
+		const frag = writer.createDocumentFragment();
+
+		Object.entries( this._data!.roots ).forEach( ( [ rootName, { content, attributes } ] ) => {
+			const parsedNodes: Array<Node | Element> = JSON.parse( content );
+			const children = [];
+
+			const rootElement = editor.model.document.getRoot( rootName )!;
+
+			for ( const [ key, value ] of attributes ) {
+				writer.setAttribute( key, value, rootElement );
+			}
+
+			for ( const child of parsedNodes ) {
+				children.push( this._createElement( writer, child ) );
+			}
+
+			frag._appendChild( children );
+
+			writer.insert( frag, rootElement );
```

Comment:
At this point, if we are creating the nodes one after another, I think that we can simply add them directly to the root (using writer). Did you encounter any problems with that?